### PR TITLE
sock: amend with zero-copy receive functions

### PR DESCRIPTION
**Pull request by miri64**
### Contribution description
This change amends the `sock` API by a set of functions to `sock` that allow provisioning of stack-internal buffers to the caller on receive.

This allows to cover two use-cases:

1. Zero-copy systems: if the stacks supported the buffer space provided by these functions can be the same that was filled in the link-layer
2. asynchronous receive within a wrapping sock layer (e.g. `sock_dtls` wrapping `sock_udp`): to receive packets of the lower level protocol asynchronously, the wrapping implementation layer would currently need to allocate its own buffer space, introducing a third buffer space in addition to the one of the application and the network stack. For a wrapping layer this is undesirable.

While there are security considerations exposing stack internal memory space to the caller, I believe they are minor, as in the end the application developer is the person in control of the node.

This change only amends the `sock` API and does not change it, so even though it is an API change, it should only require one ACK.

### Testing procedure
Only API amendments here, no functional implementation (yet). So just check if

```
make doc
```

succeeds. All `sock` tests should still compile (Murdock can do that).

### Issues/PRs references
See https://github.com/RIOT-OS/RIOT/pull/12907#discussion_r390914377

### CODE REVIEW

**Comment by miri64** [comment]
@kaspar030 ping?

**Comment by benpicco** [comment]
I suppose you want to make sure the API is sound before adding the implementation?

**Comment by miri64** [comment]
> I suppose you want to make sure the API is sound before adding the implementation?

Not just sound, but also under rough consensus. Also, since it is not just "the implementation" but several, I think its best to put those in separate PRs as we usually do with `sock`.

**Comment by benpicco** [comment]
From a first glance the parameters look fine.
Just some minor bikeshedding: wouldn't calling the functions `sock_*_recv_async()` make more sense?

**Comment by miri64** [comment]
> Just some minor bikeshedding: wouldn't calling the functions `sock_*_recv_async()` make their intention clearer?

Why? Asynchronous receive is just one use case for this (see use-cases above). Also, the name should rather reflect the functionality (provide the buffer with the receive), not the use case.

File: sys/include/net/sock/udp.h [hunk 1]
```diff
@@ -413,6 +413,44 @@ int sock_udp_get_remote(sock_udp_t *sock, sock_udp_ep_t *ep);
...
+ *
+ * @note    Function blocks if no packet is currently waiting.
+ *
+ * @return  The number of bytes received on success.
+ * @return  0, if no received data is available, but everything is in order.
+ * @return  -EADDRNOTAVAIL, if local of @p sock is not given.
+ * @return  -EAGAIN, if @p timeout is `0` and no data is available.
+ * @return  -EINVAL, if @p remote is invalid or @p sock is not properly
+ *          initialized (or closed while sock_udp_recv() blocks).
+ * @return  -ENOBUFS, if buffer space is not large enough to store received
```
Comment (leandrolanzieri) [inline sys/include/net/sock/udp.h]: Can this still be the case?

**Comment by kaspar030** [comment]
Ok, I see some difficulty: As is, the application might not free the buffer. That means the stack is forced to allocate this somewhere where it is not blocking the rest of the stack, or block itself until the buffer is freed.

In gnrc, everything is allocated anyways, so this makes sense and removes the need for another copy of the payload.

In a hypothetical single-RX-buffer stack, this is more difficult. It would probably move the allocation to within the stack.

Would it make sense to limit this to sock_async, and make the returned buffer only valid within the async callback? That would also simplify the API, as `buf_context` could be dropped.

**Comment by miri64** [comment]
> Would it make sense to limit this to sock_async, and make the returned buffer only valid within the async callback? That would also simplify the API, as `buf_context` could be dropped.

I don't think so. As the event handling for `sock_async` is not implemented in stack we would need an externally available `free` function. This would _still_ open the API for misuse.

**Comment by miri64** [comment]
> In a hypothetical single-RX-buffer stack, this is more difficult. It would probably move the allocation to within the stack.

Then for those stacks zero-copy would not be possible?

**Comment by miri64** [comment]
> As is, the application might not free the buffer.

This would be faulty usage anyway.

**Comment by kaspar030** [comment]
> This would be faulty usage anyway.

Well, not freeing at all would probably be faulty. But freeing at an arbitrary time would be valid, as proposed.

**Comment by kaspar030** [comment]
> > Would it make sense to limit this to sock_async, and make the returned buffer only valid within the async callback? That would also simplify the API, as `buf_context` could be dropped.
> 
> I don't think so. As the event handling for `sock_async` is not implemented in stack we would need an externally available `free` function. This would _still_ open the API for misuse.

True. Ok, let's go with it.

**Comment by kaspar030** [comment]
> Then for those stacks zero-copy would not be possible?

I think in the very limited stacks that fit only one packet, it is acceptable to just consider the RX buffer full as long as the application blocks it. The alternative is to have full memory, which has the same effect in practice (dropped packets).

**Comment by kaspar030** [comment]
ACK from my side, but I think @leandrolanzieri has a point.

File: sys/include/net/sock/udp.h (hunk 1 above)
Comment (miri64) [reply sys/include/net/sock/udp.h]: Mh.. not really.

**Comment by miri64** [comment]
BTW in case anybody is wondering, why there is no TCP equivalent: As TCP needs a buffer to fill anyway, I think it makes little sense to provide this functionality for this protocol.

**Comment by miri64** [comment]
Addressed @leandrolanzieri and squashed directly. I also piggy-backed some long overdue doc fixes ;-).

**Comment by miri64** [comment]
Oops, removed wrong error doc and forgot to remove it for `sock_dtls` and `sock_ip`. Now fixed!

**Comment by miri64** [comment]
Last build passed. Since then only doc was changed. There are only function declarations introduced here, but no implementations. So I think we can skip the compile tests.

**Comment by miri64** [comment]
Found yet another doc error and squashed directly

**Comment by miri64** [comment]
Travis succeeded, but it isn't showed in GitHub somehow. Let's merge this so I can go ahead with the implementation part.

**Pull request merged**